src/datastore/postgres-notifier.ts

```diff
@@ -0,0 +1,138 @@
+import { ClientConfig } from 'pg';
+import createPostgresSubscriber, { Subscriber } from 'pg-listen';
+import { logError } from '../helpers';
+import { AddressTxUpdateInfo, DbTokenMetadataQueueEntry } from './common';
+
+export type PgTxNotificationPayload = {
+  txId: string;
+};
+
+export type PgBlockNotificationPayload = {
+  blockHash: string;
+  txIds: string[];
+  microblocksAccepted: string[];
+  microblocksStreamed: string[];
+};
+
+export type PgAddressNotificationPayload = {
+  info: AddressTxUpdateInfo;
+};
+
+export type PgTokenMetadataNotificationPayload = {
+  entry: DbTokenMetadataQueueEntry;
+};
+
+export type PgNameNotificationPayload = {
+  nameInfo: string;
+};
+
+export type PgTokensNotificationPayload = {
+  contractID: string;
+};
+
+export type PgNotificationPayload =
+  | PgBlockNotificationPayload
+  | PgTxNotificationPayload
+  | PgAddressNotificationPayload
+  | PgTokenMetadataNotificationPayload
+  | PgNameNotificationPayload
+  | PgTokensNotificationPayload;
+
+export type PgNotification = {
+  type: string;
+  payload: PgNotificationPayload;
+};
+
+export type PgNotificationCallback = (notification: PgNotification) => void;
+
+/**
+ * As
+ */
+export class PgNotifier {
+  subscriber?: Subscriber;
+  eventCallback?: PgNotificationCallback;
+
+  constructor(clientConfig: ClientConfig) {
+    this.subscriber = createPostgresSubscriber(clientConfig, {
+      serialize: data =>
+        JSON.stringify(data, (_, value) => {
+          if (typeof value === 'bigint') {
+            return value.toString() + 'n';
+          }
+          if (value instanceof Map) {
+            return {
+              dataType: 'Map',
+              value: Array.from(value.entries()),
+            };
+          }
+          if (value instanceof Set) {
+            return {
+              dataType: 'Set',
+              value: Array.from(value.entries()),
+            };
+          }
+          // eslint-disable-next-line @typescript-eslint/no-unsafe-return
+          return value;
+        }),
+      parse: serialized =>
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-return
+        JSON.parse(serialized, (_, value) => {
```

Comment:
Maybe safer and more performant to require all `PgNotificationPayloads` to be plain-old-JSON? Like this
```ts
export interface AddressTxUpdateInfo {
  address: string;
  txs: Map<string, Set<AddressTxUpdateEventInfo>>;
}
```
could be refactored to
```ts
export interface AddressTxUpdateInfo {
  address: string;
  txs: Record<string, AddressTxUpdateEventInfo[]>; // Record is just typescript type for object
}
```

And `bigint` properties can instead be string-quoted integers i.e. `"12345"` as opposed to `12345n`

Reply:
Yeah that's a good idea 👍